File: js130/challenges/medium/robot-name.js
class Robot {
  static chosenNames = [];

  constructor() {
    this.robotName = this.name();
  }

  name() {
    if (this.robotName) return this.robotName;
    let randomName = this.generateRandomName();

    while(Robot.chosenNames.includes(randomName)) {
      randomName = this.generateRandomName();
    }
    Robot.chosenNames.push(randomName);
    this.robotName = randomName;
    return randomName;
  }

  generateRandomName() {
    let alphabets = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    let digits = '0123456789';
    let name = '';
    name += alphabets[Math.floor((Math.random() * alphabets.length))];
    name += alphabets[Math.floor((Math.random() * alphabets.length))];
    name += digits[Math.floor((Math.random() * digits.length))];
    name += digits[Math.floor((Math.random() * digits.length))];
    name += digits[Math.floor((Math.random() * digits.length))];
    return name;
  }

  reset() {
    this.robotName = undefined;
    this.robotName = this.name();
  }
}

module.exports = Robot;

File: py130/challenges/medium_2/robot_name.py
import random

class Robot:

    used_names = []

    def __init__(self):
        self._generate_name()

    def _generate_name(self):
        
        while True:
            name = ''
            letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
            digits = '0123456789'

            name += random.choice(letters)
            name += random.choice(letters)
            name += random.choice(digits)
            name += random.choice(digits)
            name += random.choice(digits)

            if name not in Robot.used_names:
                Robot.used_names.append(name)
                self._name = name
                break

    @property
    def name(self):
        return self._name

    def reset(self):
        self._generate_name()
